pkg/api/validation/validation.go

```diff
@@ -1937,6 +1970,58 @@ func ValidateVolumeMounts(mounts []api.VolumeMount, volumes sets.String, contain
 	return allErrs
 }
 
+func ValidateVolumeDevices(devices []api.VolumeDevice, volumes []api.Volume, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	devicepath := sets.NewString()
+	devicename := sets.NewString()
+
+	if devices != nil && !utilfeature.DefaultFeatureGate.Enabled(features.BlockVolumeSupport) {
+		allErrs = append(allErrs, field.Forbidden(fldPath.Child("volumeDevices"), "Container volumeDevices is disabled by feature-gate"))
+		return allErrs
+	}
+	// pvcvolumes := getClaimVolumesList(volumes)
+	if devices != nil {
+		for i, dev := range devices {
+			idxPath := fldPath.Index(i)
+			devName := dev.Name
+			devPath := dev.DevicePath
+			volSource := getVolumeSourceStringFromName(devName, volumes)
+			if len(devName) == 0 {
+				allErrs = append(allErrs, field.Required(idxPath.Child("name"), ""))
+			}
+			if devicename.Has(devName) {
+				allErrs = append(allErrs, field.Invalid(idxPath.Child("name"), devName, "must be unique"))
+			}
+			// Must be PersistentVolumeClaim volume source
+			// if !pvcvolumes.Has(devName) {
+			//	allErrs = append(allErrs, field.Required(idxPath.Child("name"), "volumeDevice (block devices) must have a corresponding PersistentVolumeClaim volume source type for this volume name"))
+			// }
+			// Must be PersistentVolumeClaim volume source
+			if volSource.PersistentVolumeClaim == nil {
```

Comment:
Can you distinguish between:
* Volume is not in the volumes list
* The volume in the volumes list is not a PVC

Reply:
yep, will fix up